### PR TITLE
Transaction processor refactor

### PR DESCRIPTION
**Pull request by jmederosalvarado**
## Changes

- Basically splits `TransactionProcessor.Execute` logic into multiple `virtual` methods to improve on both readability and extensibility.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

[Pre-Merge Posdao Tests passed](https://github.com/NethermindEth/nethermind/actions/runs/5243108204/jobs/9467361855)
[Post-Merge Posdao Tests passed](https://github.com/NethermindEth/nethermind/actions/runs/5243110127/jobs/9467366295)

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 1]
```diff
@@ -378,124 +405,160 @@ senderBalance < (UInt256)transaction.GasLimit * transaction.MaxFeePerGas + value
...
-                    }
-                    else
-                    {
-                        _worldState.CreateAccount(gasBeneficiary, fees);
-                    }
+            return true;
+        }
 
-                    UInt256 burntFees = !transaction.IsFree() ? (ulong)spentGas * block.BaseFeePerGas : 0;
+        protected virtual void Execute(Transaction tx, BlockHeader blk, ITxTracer tracer, ExecutionOptions opts)
```
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:443]: Minor: I would put this method as first to improve code readability, as it is calling all other methods

**Review by marcindsobczak — APPROVED** [review]
Looks correct

File: src/Nethermind/Nethermind.State/StateProvider.cs [hunk 2]
```diff
@@ -395,6 +395,18 @@ public void CreateAccount(Address address, in UInt256 balance, in UInt256 nonce)
             PushNew(address, account);
         }
 
+        public void AddToBalanceAndCreateIfNotExists(Address address, in UInt256 balance, IReleaseSpec spec)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.State/StateProvider.cs:398]: nice, common pattern. 

WithdrawalProcessor.cs will like this method as well :)  https://github.com/NethermindEth/nethermind/blob/master/src/Nethermind/Nethermind.Consensus/Withdrawals/WithdrawalProcessor.cs#L41

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 3]
```diff
@@ -93,278 +94,396 @@ public void Trace(Transaction transaction, BlockHeader block, ITxTracer txTracer
...
-                _worldState.GetNonce(tx.SenderAddress ?? Address.Zero));
+            // restore is CallAndRestore - previous call, we will restore state after the execution
+            bool restore = opts.HasFlag(ExecutionOptions.Restore);
+            // commit - is for standard execute, we will commit thee state after execution
+            // !commit - is for build up during block production, we won't commit state after each transaction to support rollbacks
+            // we commit only after all block is constructed
+            bool commit = opts.HasFlag(ExecutionOptions.Commit) || !spec.IsEip658Enabled;
 
-            if (txTracer.IsTracingReceipt)
+            if (!ValidateStatic(tx, blk, spec, tracer, opts, out long intrinsicGas))
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:110]: Cosmetic: better methods names: ValidateTx, Validate?

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 4]
```diff
@@ -93,278 +94,396 @@ public void Trace(Transaction transaction, BlockHeader block, ITxTracer txTracer
             Execute(transaction, block, txTracer, ExecutionOptions.NoValidation);
         }
 
-        private void QuickFail(Transaction tx, BlockHeader block, ITxTracer txTracer, bool eip658NotEnabled,
-            string? reason)
+        protected virtual void Execute(Transaction tx, BlockHeader blk, ITxTracer tracer, ExecutionOptions opts)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:97]: don't like blk :D for blockHeader Why not blockHeader or just header?

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 5]
```diff
@@ -93,278 +94,396 @@ public void Trace(Transaction transaction, BlockHeader block, ITxTracer txTracer
...
+                    QuickFail(tx, blk, spec, tracer, "block gas limit exceeded");
+                    return false;
                 }
             }
 
-            if (!_worldState.AccountExists(caller))
+            return true;
+        }
+
+        // TODO Should we remove this already
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:271]: proposition: to resolve this ToDo we could add _logger with warning here. and observe on nodes?

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 6]
```diff
@@ -93,278 +94,396 @@ public void Trace(Transaction transaction, BlockHeader block, ITxTracer txTracer
...
+                    QuickFail(tx, blk, spec, tracer, "insufficient sender balance");
+                    return false;
+                }
             }
 
-            // Do not charge gas if noValidation is set
-            if (!noValidation) _worldState.SubtractFromBalance(caller, senderReservedGasPayment, spec);
-            if (commit)
+            if (validate)
+                _worldState.SubtractFromBalance(tx.SenderAddress, senderReservedGasPayment, spec);
```
Comment (deffrian) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:368]: I think we can move it inside if, as senderReservedGasPayment for system txs is zero

**Review by deffrian — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs (hunk 4 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs]: done

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 7]
```diff
@@ -379,122 +498,71 @@ senderBalance < (UInt256)transaction.GasLimit * transaction.MaxFeePerGas + value
...
-                _worldState.Commit(spec, txTracer.IsTracingState ? txTracer : NullStateTracer.Instance);
-            }
+                UInt256 fees = (UInt256)spentGas * premiumPerGas;
+                UInt256 burntFees = !tx.IsFree() ? (UInt256)spentGas * blk.BaseFeePerGas : 0;
 
-            if (!noValidation && notSystemTransaction)
-            {
-                block.GasUsed += spentGas;
-            }
+                if (_worldState.AccountExists(blk.GasBeneficiary))
```
[the remark this answers is not in the document]
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs]: it wasn't on the IWorldState interface so I didn't use it. But I'll try to change it

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs (hunk 5 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:289]: done

**Review by smartprogrammer93 — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 8]
```diff
@@ -397,121 +518,84 @@ senderBalance < (UInt256)transaction.GasLimit * transaction.MaxFeePerGas + value
...
             }
 
-            if (_logger.IsTrace) _logger.Trace("Gas spent: " + spentGas);
+            if (validate && !tx.IsSystem())
+                header.GasUsed += spentGas;
+
+            return true;
+        }
+
+        protected bool PayFees(Transaction tx, BlockHeader header, IReleaseSpec spec, ITxTracer tracer, in TransactionSubstate substate, in long spentGas, in UInt256 premiumPerGas, in byte statusCode)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:556]: PayFees return true only? bool not needed here?

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs (hunk 8 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:556]: got it :) make sense :)

**Pull request merged**